### PR TITLE
Add utils/controlrefcheck.py

### PR DESCRIPTION
**Pull request by Mab879**
#### Description:

Add a script to test if the control id equals the reference of a rule. 

#### Rationale:
To ensure that the control file and the rules are in sync.

#### Review Hints:

`$ ./utils/controlrefcheck.py rhel8 cis_rhel8 cis`

Example output:

```
Skipping reload_dconf_db as it does not match a CIS id.
Skipping enable_authselect as it does not match a CIS id.
selinux_state cis@rhel8 1.7.1.4 does not match the control id 1.6.1.5
configure_crypto_policy cis@rhel8 1.10,1.11 does not match the control id 1.1
package_xinetd_removed cis@rhel8 2.1.1 does not match the control id 2.2.1
rsyslog_files_permissions cis@rhel8 4.2.3 does not match the control id 4.2.1.4
sudo_require_reauthentication cis@rhel8 5.3.6 does not match the control id 5.3.5
accounts_password_pam_retry cis@rhel8 5.4.1 does not match the control id 5.5.1
accounts_passwords_pam_faillock_unlock_time cis@rhel8 5.4.2 does not match the control id 5.5.2
set_password_hashing_algorithm_systemauth cis@rhel8 5.4.4 does not match the control id 5.5.4
set_password_hashing_algorithm_passwordauth cis@rhel8 5.4.4 does not match the control id 5.5.4
```

### CODE REVIEW

**Comment by jhrozek** [comment]
This is great! I already used the script to find bugs in OCP4 PCI-DSS (PR #10098).
One suggestion: In `get_rule_object`, could we handle `None` returned from `all_rules.get()` to handle the situation I ran into?

**Comment by jhrozek** [comment]
And a question: This seems like something that should be ran in CI after we fix the initial issues. Is that what the hunk that modifies ctest does?

**Comment by Mab879** [comment]
> This is great! I already used the script to find bugs in OCP4 PCI-DSS (PR #10098). One suggestion: In `get_rule_object`, could we handle `None` returned from `all_rules.get()` to handle the situation I ran into?

Yes, I can work on that.


> And a question: This seems like something that should be ran in CI after we fix the initial issues. Is that what the hunk that modifies ctest does?
That is correct.

**Review by jhrozek — APPROVED** [review]
LGTM
I tried running the tool against the RHEL profiles that you fixed and verified that they are indeed fixed. Running against OCP4 profiles that have issues does produce meaningful messages.

ship it :-)

File: utils/controlrefcheck.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    sys.stderr.write("Hint: run source ./.pyenv.sh\n")
+    exit(6)
+
+SSG_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
+RULES_JSON = os.path.join(SSG_ROOT, "build", "rule_dirs.json")
+BUILD_CONFIG = os.path.join(SSG_ROOT, "build", "build_config.yml")
+CONTROLS_DIR = os.path.join(SSG_ROOT, "controls")
+
+
+def parse_args() -> argparse.Namespace:
```
Comment (jan-cerny) [inline utils/controlrefcheck.py:26]: Please document the script in redthedocs.io.

File: utils/controlrefcheck.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+try:
+    from ssg.build_cpe import ProductCPEs
+    import ssg.build_profile
+    import ssg.controls
+    import ssg.environment
+    import ssg.products
+except (ModuleNotFoundError, ImportError):
+    sys.stderr.write("Unable to load ssg python modules.\n")
+    sys.stderr.write("Hint: run source ./.pyenv.sh\n")
+    exit(6)
```
[the remark this answers is not in the document]
Comment (Mab879) [reply utils/controlrefcheck.py:18]: Six was the number that I was at when I added that check. I add a unique exit code to each location that uses `exit`.

File: utils/controlrefcheck.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+def check_files(json_path: str, controls_dir: str) -> None:
+    if not os.path.exists(json_path):
+        sys.stderr.write(f"JSON at {json_path} was not found.\n")
+        sys.stderr.write("Run  ./utils/rule_dir_json.py to create.")
+        exit(3)
+    if not os.path.exists(controls_dir):
+        sys.stderr.write(f"Controls directory {controls_dir} was not found.\n")
+        exit(4)
+    if not os.path.isdir(controls_dir):
+        sys.stderr.write("Given controls directory {controls_dir} is not a directory.\n")
```
[the remark this answers is not in the document]
Comment (Mab879) [reply utils/controlrefcheck.py]: Fixed

File: utils/controlrefcheck.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+def check_cis(reference: str, control_id: str) -> bool:
+    return reference == 'cis' and not re.match(r"\d(\.\d+){0,3}", control_id)
+
+
+def check_reference(reference: str, rule_object: ssg.build_yaml.Rule, control_id: str) -> bool:
+    return reference in rule_object.references and control_id not in \
+        rule_object.references[reference].split(',')
+
+
+def main():
```
[the remark this answers is not in the document]
Comment (Mab879) [reply utils/controlrefcheck.py:117]: Do we want to lower the threshold in Code Climate, as this function passed the complexity before I changed it?

File: utils/controlrefcheck.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jan-cerny) [reply utils/controlrefcheck.py:18]: aha, thanks for exmplanation

File: utils/controlrefcheck.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jan-cerny) [reply utils/controlrefcheck.py:117]: probably not because in other situations I find Codeclimate strict enough

**Review by jan-cerny — APPROVED** [review]
I have tried to run it and I also tried to break reference  consistency in some rule.ymls and the script has shown the expected output. I have also executed the CTest locally and I have seen that it runs in the upstream GitHub Actions in the Fedora latest and other jobs.

**Pull request merged**